vertical-pod-autoscaler/pkg/crdcache/crdcache_test.go

```diff
@@ -0,0 +1,147 @@
+*/
+
+package crdcache
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	extensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	extensionclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/fake"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/utils/test"
+	dynamic "k8s.io/client-go/dynamic/fake"
+)
+
+var (
+	scalelabelFields   = []string{"status", "scaleLabelSelector"}
+	scaleLabelSelector = "." + strings.Join(scalelabelFields, ".")
+	scaleSubResource   = &extensionsv1.CustomResourceSubresources{
+		Scale: &extensionsv1.CustomResourceSubresourceScale{
+			LabelSelectorPath: &scaleLabelSelector,
+		},
+	}
+	newCrd = func(name string, subResource *extensionsv1.CustomResourceSubresources) *extensionsv1.CustomResourceDefinition {
+		return test.CustomResourceDefinition().WithName(name).WithSubresources(subResource).
+			WithGroupVersion("test.org", "v1").Get()
+	}
+	fooCrd         = newCrd("foo", scaleSubResource)
+	barCrd         = newCrd("bar", scaleSubResource)
+	notScalableCrd = newCrd("not-scalable", nil)
+
+	newUnstructured = func(name, namespace string, crd *extensionsv1.CustomResourceDefinition, scaleLabels string,
+		owners []metav1.OwnerReference) *unstructured.Unstructured {
+		return test.Unstructured().
+			WithName(name).WithNamespace(namespace).
+			WithApiVersionKind(test.CrdApiVersionAndKind(crd)).
+			AddNestedField(scaleLabels, scalelabelFields...).
+			WithOwnerReferences(owners).Get()
+	}
+
+	foo1         = newUnstructured("foo1", "ns1", fooCrd, "name=foo1", nil)
+	foo2         = newUnstructured("foo2", "ns2", fooCrd, "name=foo2", nil)
+	bar1         = newUnstructured("bar1", "ns2", barCrd, "app=bar", nil)
+	notScalable1 = newUnstructured("not-scalable1", "ns1", notScalableCrd, "", nil)
+	objNotExist1 = newUnstructured("obj-not-exist1", "ns2", fooCrd, "name=obj-not-exist1", nil)
+	objHasOwner  = newUnstructured("has-owner1", "ns1", fooCrd, "",
+		[]metav1.OwnerReference{{APIVersion: "v2", Kind: "kind1", Name: "owner"}})
+
+	fooGvr, barGvr = getGroupVersionResource(fooCrd), getGroupVersionResource(barCrd)
+	notScalableGvr = getGroupVersionResource(notScalableCrd)
+)
+
+func TestFetchSelector(t *testing.T) {
+	crdCache := newFakeCrdCache()
+	for i, testCase := range []struct {
+		gvr       schema.GroupVersionResource
+		obj       *unstructured.Unstructured
+		expectErr error
+	}{
+		{gvr: fooGvr, obj: foo1},
+		{gvr: fooGvr, obj: foo2},
+		{gvr: barGvr, obj: bar1},
+		{gvr: notScalableGvr, obj: notScalable1, expectErr: fmt.Errorf("resource %s hasn't scale sub-resource", notScalableGvr.String())},
+		{gvr: fooGvr, obj: objNotExist1, expectErr: fmt.Errorf(`foos.test.org "obj-not-exist1" not found`)},
+	} {
+		msg := fmt.Sprintf("case %d", i)
```

Comment:
Please use `t.Run` as test case wrapper.
```
t.Run(testCase.name, func(t *testing.T) {
    (...)
})
```
so adding `msg := fmt.Sprintf("case %d", i)` everywhere should not be required.